sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala

```diff
@@ -45,6 +54,84 @@ class ConnectionProviderSuite extends ConnectionProviderSuiteBase with SharedSpa
     }
   }
 
+  test("Throw an error selecting from an empty list of providers on create") {
+    val providerBase = new ConnectionProviderBase() {
+      override val providers = Seq.empty
+    }
+
+    val err1 = intercept[IllegalArgumentException] {
+      providerBase.create(mock[Driver], Map.empty, None)
+    }
+    assert(err1.getMessage.contains("Empty list of JDBC connection providers"))
+
+    val err2 = intercept[IllegalArgumentException] {
+      providerBase.create(mock[Driver], Map.empty, Some("test"))
+    }
+    assert(err2.getMessage.contains("Empty list of JDBC connection providers"))
+  }
+
+  test("Throw an error when more than one provider is available on create") {
+    val provider1 = new JdbcConnectionProvider() {
```

Comment:
I have the feeling that `TestJdbcConnectionProvider(val name, val getConnectionImpl...)` class can be created in a common area. Such way several copy-paste can be eliminated, right?

Reply:
IMHO, I would have to provide closures for `canHandle` and `getConnection` to the class anyway, so I am not quite sure what the test wrapper class would accomplish.